### ISSUE TITLE
Add an edge_bfs traversal to match edge_dfs

### ISSUE THREAD

**Issue by h-vetinari** [networkx/networkx#2792]
`networkx.algorithms.traversal.bfs_edges` does not correctly return all the edges of a graph if a node has more than one in-edge. My expectation would be (and critical in my application) is that in a traversal, _all_ edges in the graph are returned at some point in the iteration.

Here's an illustration

```
import networkx as nx
from networkx.algorithms.traversal import bfs_edges
from random import shuffle

G = nx.DiGraph()
e_in = [(1, 2), (1, 3), (2, 4), (2, 5), (3, 6), (3, 7), (4, 8), (6, 8), (5, 9), (7, 9), (8,10), (9,10)]
# shuffle(e_in) # test that bfs_edges generates correct order
G.add_edges_from(e_in)
e_out = [e for e in bfs_edges(G, 1)]
print('Received: ', e_in)
print('Returned: ', e_out)
print('Set difference: ', set(e_in) ^ set(e_out))
# Set difference:  {(6, 8), (9, 10), (7, 9)}

# visualisation
# import graphviz as gv
# gv.Source(nx.drawing.nx_pydot.to_pydot(G).to_string())
```

The issue is in `generic_bfs_edges` in https://networkx.github.io/documentation/latest/_modules/networkx/algorithms/traversal/breadth_first_search.html, where a node gets excluded from forming more edge pairs if it gets visited once - hence the issue if a node has more than one in-edge. However, it's not a solution to completely exclude this visitor-mechanism, because then the respective child-nodes get added to the queue multiple times.

```
from collections import deque, Counter
def generic_bfs_edges_no_visitors(G, source, neighbors=None):
    visited = {source}
    queue = deque([(source, neighbors(source))])
    while queue:
        parent, children = queue[0]
        try:
            child = next(children)
            if child not in visited:
                yield parent, child
                # visited.add(child)
                queue.append((child, neighbors(child)))
        except StopIteration:
            queue.popleft()

e_out = [e for e in generic_bfs_edges_no_visitors(G, 1, G.successors)]
print('Received: ', e_in)
print('Returned: ', e_out)
print('Set difference: ', set(e_in) ^ set(e_out))
# Set difference:  set()
print('Multiset difference: ', Counter(e_out) - Counter(e_in))
# Multiset difference:  Counter({(8, 10): 1, (9, 10): 1})
```

Here's my proposed fix. I don't know enough about the rest of the library to see if this would break anything, but getting graph traversal right sounds pretty fundamental to me.

```
def generic_bfs_edges_proposed_fix(G, source, neighbors=None):
    visited = {source}
    queue = deque([(source, neighbors(source))])
    while queue:
        parent, children = queue[0]
        try:
            child = next(children)
            yield parent, child # moved this line out of the if-condition!
            if child not in visited:
                visited.add(child)
                queue.append((child, neighbors(child)))
        except StopIteration:
            queue.popleft()

e_out = [e for e in generic_bfs_edges_proposed_fix(G, 1, G.successors)]
print('Received: ', e_in)
print('Returned: ', e_out)
print('Set difference: ', set(e_in) ^ set(e_out))
# Set difference:  set()
print('Multiset difference: ', Counter(e_out) - Counter(e_in))
# Multiset difference:  Counter()
```

**Comment by jfinkels** [comment]
I think the `bfs_edges()` function yields the edges underlying a breadth-first search of the nodes of a graph. What you are requesting sounds like it might be a breadth-first search on the edges of the graph, analagous to the depth-first search on the edges of the graph provided by the `edge_dfs` function: https://networkx.github.io/documentation/latest/reference/algorithms/generated/networkx.algorithms.traversal.edgedfs.edge_dfs.html#networkx.algorithms.traversal.edgedfs.edge_dfs. If so, then that function does not yet exist.

**Comment by h-vetinari** [comment]
@jfinkels Thanks for the answer. I think the function names (and the docs, to a degree) are confusing on this. Your answer that `bfs_edges` iterates over the `nodes` makes perfect sense, but the name suggests an edge traversal, and [the docs](https://networkx.github.io/documentation/stable/reference/algorithms/traversal.html) state as well that

> `bfs_edges(G, source[, reverse])` : Iterate over edges in a breadth-first-search starting at source

I think the example I showed contradicts at least this particular wording. In any case, I really needed a funtion that might be called `edge_bfs` (consistent with the naming for `dfs`-functions), so I wrote the following:

```
from collections import deque
def edge_bfs(G, source = None, neighbors = None):
    """Iterate over edges in a breadth-first search. Not implemented in networkx (as of ver 2.0)
    Accepts a single node or list of nodes as "source", and will return all edges in the graph if "source" is None."""
    if neighbors is None:
        neighbors = G.neighbors
    
    if source is None:
        source = [x for x in G.nodes if G.in_degree(x) == 0 and G.out_degree(x) > 0]
    elif source in G: # a single key
        source = [source]
    # otherwise, assume "source" is iterable of nodes
    
    source = sorted(source, key = lambda x: max([len(v) for k, v in nx.shortest_path(G, x).items()]), reverse = True)
    visited = set(source)
    queue = deque([(s, neighbors(s)) for s in source])
    while queue:
        parent, children = queue[0]
        try:
            child = next(children)
            yield parent, child
            if child not in visited:
                visited.add(child)
                queue.append((child, neighbors(child)))
        except StopIteration:
            queue.popleft()
```

The sorting makes sure that we never traverse an edge whose starting node has not yet been reached by all edges leading into it (which is crucial to a breadth-first search, at least in my application).

Clearly, some more API functionality is missing (e.g. `orientation = 'original'|'reverse'|'ignore'` instead of `neighbors = None`), but I wanted to see if there is any interest at all first. If yes, should we rename this issue to a proposal for `edge_bfs`?

**Comment by h-vetinari** [comment]
Still haven't found the time for a PR, but here's my code for a working `bfs_edges`, minus some of the API-functionality that might be desirable.

The issue with the code I posted above is that sorting by descending `nx.shortest_path` can be wrong if there are connections of different lengths between two nodes. The version below needs to calculate the longest path for each subgraph starting at a given node, but is provably correct (assuming the edge about to be returned starts at a node with a predecessor that hasn't been encountered yet, that predecessor would have a longer `dag_longest_path` and therefore be ranked before the current one in the priority queue - a contradiction).

As a side note, `dag_longest_path` -- like `shortest_path` should have a `source` parameter. This would save me having to construct the subgraph explicitly in the priority of the queue(d element).

```
import heapq

from networkx import DiGraph
from networkx.algorithms.dag import dag_longest_path as dag_lp
from networkx.algorithms.traversal import edge_dfs


# Adapted from https://www.safaribooksonline.com/library/view/python-cookbook-3rd/9781449357337/ch01s05.html
class PriorityQueue:
    '''
    Queue-like class with push/pop that always returns the item
    with the highest priority (as specified when pushing the item).
    '''
    def __init__(self):
        self._queue = []
        self._index = 0

    def push(self, item, priority):
        heapq.heappush(self._queue, (-priority, self._index, item))
        self._index += 1
    
    def pop(self):
        return heapq.heappop(self._queue)[2]
    
    def __bool__(self):
        return bool(self._queue)
    
    def __iter__(self):
        yield from [x[2] for x in self._queue]
    
    def __getitem__(self, key):
        return self._queue[key][2]

def edge_bfs(G, source = None, neighbors = None):
    '''
    Iterate over edges of graph in a breadth-first search.
    
    Args:
        G (:class:`nx.DiGraph`):
            The graph we want to traverse. Must be directed, as well as
            acyclic.
        source (optional):
            The nodes in the graph from which to start the
            breadth-first edge traversal. The following types of
            arguments are accepted:
            
            - :obj:`None` (the default): in this case all nodes having
              an :meth:`in_degree` of 0 and an :meth:`out_degree` > 0
              are used as :attr:`source`.
            - a valid key of :attr:`G`: in this case the corresponding
              node of :attr:`G` will be the sole starting point.
            - an iterable of valid keys of :attr:`G`: in this case all
              nodes corresponding to the keys in the iterable will be
              used as :attr:`source`.
              
        neighbors (optional):
            Defaults to :obj:`None`, in which case the default
            neighbors of :attr:`G` are used.
    
    Yields:
        A tuple of two node keys, in the order of the direction of the
        edge, and continuing like that through the graph (starting from
        :attr:`source`) in a breadth-first manner.
        
    Note:
        This should reasonably be part of the :mod:`networkx`-library,
        but is not implemented (as of ver 2.1).
    '''
    if neighbors is None:
        neighbors = G.neighbors
    
    if source is None:
        source = [x for x in G.nodes if G.in_degree(x) == 0 and G.out_degree(x) > 0]
    elif source in G: # a single key
        source = [source]
    else: # otherwise, assume "source" is iterable of nodes
        source = list(source)
    
    visited = set(source)
    
    depth = lambda node_key: len(dag_lp(DiGraph(edge_dfs(G, node_key)))) \
                 if G.out_degree(node_key) > 0 else 0
    queue = PriorityQueue()
    for s in source:
        queue.push((s, neighbors(s)), depth(s))
        
    while queue:
        parent, children = queue[0]
        try:
            child = next(children)
            yield parent, child
            if child not in visited:
                visited.add(child)
                queue.push((child, neighbors(child)), depth(child))
        except StopIteration:
            queue.pop())
```
